### PR TITLE
Signup: Convert lib/signup/cart to TypeScript

### LINKED ISSUE #53128: Refactor lib/signup/cart to TS
`lib/signup/cart` is one of a couple places in calypso that accesses the shopping-cart endpoint directly without going through the `@automattic/shopping-cart` package. However, this means that it is subject to bugs that would otherwise be prevented by that package. For example, https://github.com/Automattic/wp-calypso/pull/53093

https://github.com/Automattic/wp-calypso/pull/52355 added unit tests for those functions, but that assumes the functions are being called correctly (in the above case they were not).

If we at least convert that code to TypeScript, it will help prevent future bugs.

### PR DESCRIPTION
**Pull request by sirbrillig**
#### Changes proposed in this Pull Request

`lib/signup/cart` is one of a couple places in calypso that accesses the shopping-cart endpoint directly without going through the `@automattic/shopping-cart` package. However, this means that it is subject to bugs that would otherwise be prevented by that package. For example, https://github.com/Automattic/wp-calypso/pull/53093

https://github.com/Automattic/wp-calypso/pull/52355 added unit tests for those functions, but that assumes the functions are being called correctly (in the above case they were not).

This PR converts that file to TypeScript using the types from the `@automattic/shopping-cart` package which should help make sure that they are called correctly in the future. Of course, this requires that the [calling code](https://github.com/Automattic/wp-calypso/blob/72db21758cdf71a3e4fe139c3c05090e421f3fe4/client/lib/signup/step-actions/index.js#L51) is also converted to TypeScript, but converting signup's `step-actions` is not going to be fast or easy. At least this helps move us in the right direction, and some IDEs may take advantage of it.

Fixes https://github.com/Automattic/wp-calypso/issues/53128

#### Testing instructions

This should not change any behavior. The functions are already covered by unit tests, so there should be nothing to test manually.

### CODE REVIEW

**Review by nbloomf — APPROVED** [review]
Looks good. Just for fun I also tested a checkout from signup and saw no problems.

File: client/lib/signup/cart.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	createCart: function (
+		cartKey: string,
+		newCartItems: RequestCartProduct[],
+		callback: ErrorCallback
+	): void {
+		let newCart = getEmptyResponseCart();
+		newCart = {
+			...newCart,
+			cart_key: cartKey,
+		};
```
Comment (pottedmeat) [inline client/lib/signup/cart.ts]: Could you do this instead?

```ts
const newCart = {
	...getEmptyResponseCart(),
	cart_key: cartKey,
};
```

File: client/lib/signup/cart.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+		const updatedCart: RequestCart = addProductsToCart( newCart, newCartItems );
```
Comment (pottedmeat) [inline client/lib/signup/cart.ts]: I don't think you need to type this variable.

**Review by pottedmeat — APPROVED** [review]
Looks good. Minor comments.

File: client/lib/signup/cart.ts (hunk 1 above)
Comment (sirbrillig) [reply client/lib/signup/cart.ts]: Yeah, you're right. 👍

File: client/lib/signup/cart.ts (hunk 2 above)
Comment (sirbrillig) [reply client/lib/signup/cart.ts]: Yeah, good point. I've removed that too. Both of these were relics of how I was updating the old code bit by bit.

**Pull request merged**